File: serialization/dotnet/text/src/TextSerializationWriter.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.Xml;
using Microsoft.Kiota.Abstractions;
using Microsoft.Kiota.Abstractions.Serialization;
using System.Text;
using Microsoft.Kiota.Abstractions.Extensions;
using System.Linq;

namespace Microsoft.Kiota.Serialization.Text;

/// <summary>
/// The <see cref="ISerializationWriter"/> implementation for text content types.
/// </summary>
public class TextSerializationWriter : ISerializationWriter, IDisposable {
    private readonly MemoryStream _stream = new MemoryStream();
    private readonly StreamWriter writer;
    /// <summary>
    /// Initializes a new instance of the <see cref="TextSerializationWriter"/> class.
    /// </summary>
    public TextSerializationWriter()
    {
        writer = new(_stream);
    }
    private bool written;
    /// <inheritdoc />
    public Action<IParsable> OnBeforeObjectSerialization { get; set; }
    /// <inheritdoc />
    public Action<IParsable> OnAfterObjectSerialization { get; set; }
    /// <inheritdoc />
    public Action<IParsable, ISerializationWriter> OnStartObjectSerialization { get; set; }
    /// <inheritdoc />
    public void Dispose()
    {
        _stream?.Dispose();
        writer?.Dispose();
        GC.SuppressFinalize(this);
    }
    /// <inheritdoc />
    public Stream GetSerializedContent() {
        writer.Flush();
        return _stream;
    }
    /// <inheritdoc />
    public void WriteAdditionalData(IDictionary<string, object> value) => throw new InvalidOperationException(TextParseNode.NoStructuredDataMessage);
    /// <inheritdoc />
    public void WriteBoolValue(string key, bool? value) => WriteStringValue(key, value?.ToString());
    /// <inheritdoc />
    public void WriteByteArrayValue(string key, byte[] value) => WriteStringValue(key, value.Any() ? Convert.ToBase64String(value) : string.Empty);
    /// <inheritdoc />
    public void WriteByteValue(string key, byte? value) => WriteStringValue(key, value?.ToString());
    /// <inheritdoc />
    public void WriteCollectionOfObjectValues<T>(string key, IEnumerable<T> values) where T : IParsable => throw new InvalidOperationException(TextParseNode.NoStructuredDataMessage);
    /// <inheritdoc />
    public void WriteCollectionOfPrimitiveValues<T>(string key, IEnumerable<T> values) => throw new InvalidOperationException(TextParseNode.NoStructuredDataMessage);
    /// <inheritdoc />
    public void WriteDateTimeOffsetValue(string key, DateTimeOffset? value) => WriteStringValue(key, value.HasValue ? value.Value.ToString() : null);
    /// <inheritdoc />
    public void WriteDateValue(string key, Date? value) => WriteStringValue(key, value?.ToString());
    /// <inheritdoc />
    public void WriteDecimalValue(string key, decimal? value) => WriteStringValue(key, value?.ToString());
    /// <inheritdoc />
    public void WriteDoubleValue(string key, double? value) => WriteStringValue(key, value?.ToString());
    /// <inheritdoc />
    public void WriteFloatValue(string key, float? value) => WriteStringValue(key, value?.ToString());
    /// <inheritdoc />
    public void WriteGuidValue(string key, Guid? value) => WriteStringValue(key, value?.ToString());
    /// <inheritdoc />
    public void WriteIntValue(string key, int? value) => WriteStringValue(key, value?.ToString());
    /// <inheritdoc />
    public void WriteLongValue(string key, long? value) => WriteStringValue(key, value?.ToString());
    /// <inheritdoc />
    public void WriteNullValue(string key) => WriteStringValue(key, "null");
    /// <inheritdoc />
    public void WriteObjectValue<T>(string key, T value) where T : IParsable => throw new InvalidOperationException(TextParseNode.NoStructuredDataMessage);
    /// <inheritdoc />
    public void WriteSbyteValue(string key, sbyte? value) => WriteStringValue(key, value?.ToString());
    /// <inheritdoc />
    public void WriteStringValue(string key, string value)
    {
        if(!string.IsNullOrEmpty(key))
            throw new InvalidOperationException(TextParseNode.NoStructuredDataMessage);
        if(!string.IsNullOrEmpty(value))
            if(written)
                throw new InvalidOperationException("a value was already written for this serialization writer, text content only supports a single value");
            else {
                writer.Write(value);
                written = true;
            }
    }
    /// <inheritdoc />
    public void WriteTimeSpanValue(string key, TimeSpan? value) => WriteStringValue(key, value.HasValue ? XmlConvert.ToString(value.Value) : null);
    /// <inheritdoc />
    public void WriteTimeValue(string key, Time? value) => WriteStringValue(key, value?.ToString());
    /// <inheritdoc />
    void ISerializationWriter.WriteCollectionOfEnumValues<T>(string key, IEnumerable<T?> values) => throw new InvalidOperationException(TextParseNode.NoStructuredDataMessage);
    /// <inheritdoc />
    void ISerializationWriter.WriteEnumValue<T>(string key, T? value) => WriteStringValue(key, value.HasValue ? value.Value.ToString().ToFirstCharacterLowerCase() : null);
}

File: serialization/typescript/text/src/textSerializationWriter.ts
import {
  DateOnly,
  Duration,
  Parsable,
  SerializationWriter,
  TimeOnly,
} from "@microsoft/kiota-abstractions";

export class TextSerializationWriter implements SerializationWriter {
  private static noStructuredDataMessage =
    "text does not support structured data";
  private readonly writer: string[] = [];
  public onBeforeObjectSerialization: ((value: Parsable) => void) | undefined;
  public onAfterObjectSerialization: ((value: Parsable) => void) | undefined;
  public onStartObjectSerialization:
    | ((value: Parsable, writer: SerializationWriter) => void)
    | undefined;
  public writeStringValue = (key?: string, value?: string): void => {
    if (key || key !== "") {
      throw new Error(TextSerializationWriter.noStructuredDataMessage);
    }
    if (value) {
      if (this.writer.length > 0) {
        throw new Error(
          "a value was already written for this serialization writer, text content only supports a single value"
        );
      } else {
        this.writer.push(value);
      }
    }
  };
  public writeBooleanValue = (key?: string, value?: boolean): void => {
    if (value) {
      this.writeStringValue(key, `${value}`);
    }
  };
  public writeNumberValue = (key?: string, value?: number): void => {
    if (value) {
      this.writeStringValue(key, `${value}`);
    }
  };
  public writeGuidValue = (key?: string, value?: string): void => {
    if (value) {
      this.writeStringValue(key, `"${value}"`);
    }
  };
  public writeDateValue = (key?: string, value?: Date): void => {
    if (value) {
      this.writeStringValue(key, `"${value.toISOString()}"`);
    }
  };
  public writeDateOnlyValue = (key?: string, value?: DateOnly): void => {
    if (value) {
      this.writeStringValue(key, `"${value.toString()}"`);
    }
  };
  public writeTimeOnlyValue = (key?: string, value?: TimeOnly): void => {
    if (value) {
      this.writeStringValue(key, `"${value.toString()}"`);
    }
  };
  public writeDurationValue = (key?: string, value?: Duration): void => {
    if (value) {
      this.writeStringValue(key, `"${value.toString()}"`);
    }
  };
  public writeNullValue = (key?: string): void => {
    this.writeStringValue(key, `null`);
  };
  public writeCollectionOfPrimitiveValues = <T>(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    key?: string,
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    values?: T[]
  ): void => {
    throw new Error(TextSerializationWriter.noStructuredDataMessage);
  };
  public writeCollectionOfObjectValues = <T extends Parsable>(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    key?: string,
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    values?: T[]
  ): void => {
    throw new Error(TextSerializationWriter.noStructuredDataMessage);
  };
  public writeObjectValue = <T extends Parsable>(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    key?: string,
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    value?: T
  ): void => {
    throw new Error(TextSerializationWriter.noStructuredDataMessage);
  };
  public writeEnumValue = <T>(
    key?: string | undefined,
    ...values: (T | undefined)[]
  ): void => {
    if (values.length > 0) {
      const rawValues = values
        .filter((x) => x !== undefined)
        .map((x) => `${x}`);
      if (rawValues.length > 0) {
        this.writeStringValue(
          key,
          rawValues.reduce((x, y) => `${x}, ${y}`)
        );
      }
    }
  };
  public getSerializedContent = (): ArrayBuffer => {
    return this.convertStringToArrayBuffer(this.writer.join(``));
  };

  private convertStringToArrayBuffer = (str: string): ArrayBuffer => {
    const arrayBuffer = new ArrayBuffer(str.length);
    const uint8Array = new Uint8Array(arrayBuffer);
    for (let i = 0; i < str.length; i++) {
      uint8Array[i] = str.charCodeAt(i);
    }
    return arrayBuffer;
  };

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  public writeAdditionalData = (value: Record<string, unknown>): void => {
    throw new Error(TextSerializationWriter.noStructuredDataMessage);
  };
}
